refactor agent pipeline reloading to avoid double live pipelines with same settings

colinsurprenant opened the pull request:
this fixes #6605 

- refactor agent reload sequence to make sure a pipeline is shutdown before a new one is started with same settings.

- the goal here as stated in https://github.com/elastic/logstash/issues/6605#issuecomment-276751216 is to introduce the minimum changes to fix the dual pipeline problem for the next minor release, knowing that @ph is working on a deeper refactor of the whole config change management.

colinsurprenant commented:
We just discussed with @jordansissel to try to find a solution to get the plugins validation back before the shutdown + start sequence. Looking at it.

colinsurprenant commented on logstash-core/lib/logstash/agent.rb:
```diff
@@ -232,62 +229,126 @@ def collect_metrics?
-    new_pipeline = create_pipeline(old_pipeline.settings, new_config)
+    # check if this pipeline is not reloadable. it should not happen as per the check below
+    # but keep it here as a safety net if a reloadable pipeline was releoaded with a non reloadable pipeline
+    if !old_pipeline.reloadable?
```
it should not happen with the code as it is now, but if someone else in the future decide to trigger `update_pipeline` outside this sequence it will prevent it. it is really a defensive check.  I  don't mind removing it but at the same time it is harmless and will prevent it if we end up in a situation were a pipeline was not reloadable. more like an assertion in the code.

colinsurprenant commented on logstash-core/lib/logstash/agent.rb:
```diff
@@ -232,62 +229,126 @@ def collect_metrics?
+    # check if the new pipeline will be reloadable in which case we want to log that as an error and abort
+    if !pipeline_validator.reloadable?
+      @logger.error(I18n.t("logstash.agent.non_reloadable_config_reload"), :pipeline_id => id, :plugins => pipeline_validator.non_reloadable_plugins.map(&:class))
+      # TODO: in the original code the failure metrics were not incremented, should we do it here?
```
yeah - I wanted to avoid changing the behaviour since all tests are passing but I can add that  and a test for it I guess.

colinsurprenant replied:
actually - I think I'd prefer to leave it as is since it will change reporting behaviour, it is not tested for and it is now created to handle exception this is not the result of a exception. I suggest we create a separate issue for this.

jsvd replied:
fair enough, +1 on creating an issue to track it

colinsurprenant commented on logstash-core/lib/logstash/agent.rb:
```diff
@@ -232,62 +229,126 @@ def collect_metrics?
+    # start new pipeline now that the old one is shutdown
+    new_pipeline = create_pipeline(settings, new_config)
+    if new_pipeline.nil?
+      # this is a scenario where the configuration is valid (compilable) but the new pipeline refused to start
```
ah yeah, it was more a comment for the following section below. will clarify.

colinsurprenant commented:
@jsvd Thanks. I pushed a commit to address your suggestions. Let me know.

ph commented on logstash-core/lib/logstash/agent.rb:
```diff
@@ -232,62 +229,129 @@ def collect_metrics?
       rescue => e
         @logger.error("failed to fetch pipeline configuration", :message => e.message)
-        return
+        return nil
```
explicit return!

colinsurprenant replied:
see comment below

ph commented on logstash-core/lib/logstash/agent.rb:
```diff
@@ -232,62 +229,129 @@ def collect_metrics?
+      # and at this point NO pipeline is running
+      @logger.error("failed to create the reloaded pipeline and no pipeline is currently running", :pipeline => pipeline_id)
+      return
+    end
```
This change make sense here, I've did the same change in behavior in my refactor because of the simplicity,

ph commented on logstash-core/lib/logstash/agent.rb:
```diff
@@ -232,62 +229,129 @@ def collect_metrics?
+      @logger.error(I18n.t("logstash.agent.non_reloadable_config_reload"), :pipeline_id => id, :plugins => pipeline_validator.non_reloadable_plugins.map(&:class))
+      # TODO: in the original code the failure metrics were not incremented, should we do it here?
+      # increment_reload_failures_metrics(id, new_config, e)
+      return
```
I think we should increment reload failure, because the current state is not consistent to what we are expecting.

colinsurprenant replied:
here's what I said to @jsvd same comment: «I think I'd prefer to leave it as is since it will change reporting behaviour, it is not tested for and it is now created to handle exception this is not the result of a exception. I suggest we create a separate issue for this.»

ph replied:
I will raise that issue again with the refactor I did, we can decide when it land.

jsvd commented:
functionality LGTM, testing several corner cases revealed no bugs, both on the pipeline reload side and the metrics side.

I'd still like to remove some of the code style changes, like https://github.com/elastic/logstash/pull/6631/files#r99300484 and a few explicit `return nil` (not consistent with the rest of the code)

colinsurprenant commented:
@jsvd  thanks. the explicit `return nil` is absolutely not inconsistent and I will leave it as is. all the methods where a bare `return` is done is because these methods don't return a usable value. the `create_pipeline` methods returns either a `Pipeline` object or `nil` if it fails. This makes is explicit.  I added a comment to document the return value.

colinsurprenant commented:
Also, it's funny, I was expecting to send the crowd into a frenzy, all cheering for my changes since the previous formatting was so painful to look at. :man_shrugging:

ph commented:
@colinsurprenant well, I am still stuck in a 120 chars wide, editor.

colinsurprenant commented:
@ph time to upgrade. 2017.

ph commented on logstash-core/lib/logstash/agent.rb:
```diff
@@ -232,62 +229,130 @@ def collect_metrics?
+    else
+      @logger.error("fetched an invalid config", :config => config, :reason => exception.message)
+    end
+  end
```
+1 I did something similar.

ph commented on logstash-core/lib/logstash/agent.rb:
```diff
@@ -232,62 +229,130 @@ def collect_metrics?
+    end
 
-    return if new_pipeline.nil?
+    # we know configis valid so we are fairly comfortable to first stop old pipeline and then start new one
```
missing space between *config* and *is*.

colinsurprenant commented:
thanks @jsvd and @ph - rebased + squashed, merged in master and 5.x
